src/components/media-views.js

```diff
@@ -837,3 +846,150 @@ AFRAME.registerComponent("media-image", {
+      const ph = viewport.height;
+
+      if (!this.canvas || this.canvas.width !== pw || this.canvas.height !== ph) {
+        if (this.canvas) {
+          const oldCacheKey = `${this.el.object3D.uuid}_${this.canvas.width}_${this.canvas.height}`;
+
+          if (this.mesh.map !== errorTexture && textureCache.has(oldCacheKey)) {
+            textureCache.release(oldCacheKey);
+            this.currentPageTextureIsRetained = false;
+          }
+        }
+
+        this.canvas = document.createElement("canvas");
+        this.canvasContext = this.canvas.getContext("2d");
+        this.canvas.width = pw;
+        this.canvas.height = ph;
+      }
+
+      const renderingSrc = this.data.src;
+      const renderingIndex = index;
+
+      if (this.renderTask) {
+        // If a previous update is rendering, wait for it to finish.
+        await new Promise(resolve => {
+          const interval = setInterval(() => {
+            if (!this.renderTask) {
+              clearInterval(interval);
+              resolve();
+            }
+          }, 100);
+        });
+      }
+
+      this.renderTask = page.render({ canvasContext: this.canvasContext, viewport });
+      await this.renderTask.promise;
+      this.renderTask = null;
+
+      if (this.data.src !== renderingSrc || this.data.index !== renderingIndex) return;
+
+      const cacheKey = this.currentTextureCacheKey();
+      let cacheItem;
+      if (textureCache.has(cacheKey)) {
+        if (this.currentPageTextureIsRetained) {
+          cacheItem = textureCache.get(cacheKey);
+        } else {
+          cacheItem = textureCache.retain(cacheKey);
+        }
+      } else {
+        if (src === "error") {
+          cacheItem = errorCacheItem;
+        } else {
+          const texture = new THREE.CanvasTexture(this.canvas);
+          texture.encoding = THREE.sRGBEncoding;
+          texture.minFilter = THREE.LinearFilter;
+          cacheItem = textureCache.set(cacheKey, texture);
+        }
+      }
+
+      texture = cacheItem.texture;
+      ratio = cacheItem.ratio;
+
+      this.currentPageTextureIsRetained = true;
+    } catch (e) {
+      console.error("Error loading image", this.data.src, e);
+      texture = errorTexture;
+      this.currentPageTextureIsRetained = false;
+    }
+
+    if (!this.mesh) {
+      const material = new THREE.MeshBasicMaterial();
+      const geometry = new THREE.PlaneBufferGeometry(1, 1, 1, 1, texture.flipY);
+      material.side = THREE.DoubleSide;
+
+      this.mesh = new THREE.Mesh(geometry, material);
+      this.el.setObject3D("mesh", this.mesh);
+    }
+
+    this.mesh.material.transparent = texture == errorTexture;
+    this.mesh.material.map = texture;
+    this.mesh.material.map.needsUpdate = true;
+    this.mesh.material.needsUpdate = true;
+
+    scaleToAspectRatio(this.el, ratio);
+
+    if (texture !== errorTexture && this.data.batch) {
+      this.el.sceneEl.systems["hubs-systems"].batchManagerSystem.addObject(this.mesh);
```

Comment:
Hard to say if we should bother batching PDF pages. Feels a bit of a shame to bother atlasing the textures while quickly paging through a document, but maybe its fine.